### DISCUSSION TITLE
Extracting Bezier curves including operators

### DISCUSSION THREAD

**Discussion by petermr** [jsvine/pdfplumber#892] [Q&A]
I'm interested in extracting full information for Bezier curves (not just coordinates). AFAICS `LTCurve` does not hold Bezier operators (`M[ove]`, `C[ubic]`, `Q[uadratic]` and `Z[Close]`). I assume that the points are a mixtures of points on the curve and control points. However without the operators it is not possible to draw the curves.  

I don't know the PDF primitives but In have used their translation (e.g. in `PDFBox`) from which I can generate SVG. 

Here's an example 
https://upload.wikimedia.org/wikipedia/commons/d/d9/Circle_and_cubic_bezier.svg
and its source:
view-source:https://upload.wikimedia.org/wikipedia/commons/d/d9/Circle_and_cubic_bezier.svg
of how to generate a (good) approximation to a circle

```
<path d="M 16 128 C 16 66 66 16 128 16 S 240 66 240 128 190 240 128 240 16 190 16 128" stroke="red" stroke-width="4" fill="none"/>
 ```
This creates a 90deg arc (Move + cubic Bezier) , which is then multiplied by 4
Essentially we need The M/C/S etc. operators

I am not very familiar with `pdfminer.six` but its `LTCurve` doesn't seem to have the operators. Maybe there are implicit rules...

Here's .../pdfminer/layout.py
```
class LTCurve(LTComponent):
    """A generic Bezier curve"""


    def __init__(
        self,
        linewidth: float,
        pts: List[Point],
        stroke: bool = False,
        fill: bool = False,
        evenodd: bool = False,
        stroking_color: Optional[Color] = None,
        non_stroking_color: Optional[Color] = None,
    ) -> None:
        LTComponent.__init__(self, get_bound(pts))
        self.pts = pts
        self.linewidth = linewidth
        self.stroke = stroke
        self.fill = fill
        self.evenodd = evenodd
        self.stroking_color = stroking_color
        self.non_stroking_color = non_stroking_color

```

--

**Comment by petermr** [comment]
Here's the comparison of the full Bezier curve and the join-the-points

## original
![Screenshot 2023-05-26 at 19 01 09](https://github.com/jsvine/pdfplumber/assets/733515/ea431ebd-f46c-471b-a0f1-31ded6cd68a4)

## points polygon


<img width="407" alt="Screenshot 2023-05-28 at 14 43 18" src="https://github.com/jsvine/pdfplumber/assets/733515/ee88234d-a80b-47fa-b664-232f7c438e35">

**Comment by jsvine** [comment]
Some good news — it looks like support for Bezier curve control points has been added to `pdfminer.six` and will be available in its next release: https://github.com/pdfminer/pdfminer.six/issues/672 / https://github.com/pdfminer/pdfminer.six/blob/master/CHANGELOG.md

Once that release is published, I'll see if I can pull that info into the `pdfplumber`'s curve objects.

Is there a particular way you'd like to see those control points represented?

**Reply by petermr** [reply]
Thanks! Great news
I'm not an expert on PDF so don't know the format or data structure. I find https://math.stackexchange.com/questions/3000635/rational-bezier-curves-and-pdfs but this isn't much help. 

I'd go with the SVG format (a mixture of points and single letters (M L V H C S Q T A Z)) - it's concise, very widely used - in every browser. See https://www.w3.org/TR/SVG/paths.html .  I don't know what other vector graphics formats  do. 

Do` pdfminer.six` have a draft spec for their output? If so will be useful to see what they do. Happy to discuss it if it helps.

I assume that most users of `pdfplumber` will mainly wish to replot it , in which case SVG is a natural and easy output.  (and the few data extractors like me will probably work happily with SVG input.)

**Reply by jsvine** [reply]
> Do `pdfminer.six` have a draft spec for their output? If so will be useful to see what they do. Happy to discuss it if it helps.

It doesn't seem so officially, but the tests in the [relevant commit](https://github.com/pdfminer/pdfminer.six/commit/5114acdda61205009221ce4ebf2c68c144fc4ee5) seem to suggest it will look like this, in raw form:

```python
[
  ("m", (72.41, 433.89)),
  ("c", (72.41, 434.45), (71.96, 434.89), (71.41, 434.89)),
]
```

> I'd go with the SVG format (a mixture of points and single letters (M L V H C S Q T A Z)) - it's concise, very widely used - in every browser. 

I like this idea — thanks! I might default to `pdfminer.six`'s more list/tuple-based representation above, but could absolutely see adding a utility function `curve_to_svg_fmt(...)` to produce something that.

**Comment by petermr** [comment]
Great,
I'm happy with the LISP/nested-tuples format shown. It's easy to convert to SVG.

More generally, I am using HTML and SVG to **model** the semantics of the PDF.  In effect to recreate the author's data model.  Clearly a lot of this community are extracting from tables, and even there HTML will help a lot, especially with more complex tables (beyond the cartesians rows*cols). HTML and SVG can hold intermediate outputs which can be further analysed downstream.

Thanks to the `pdfminer.six` developers.

**Comment by QuentinAndre11** [comment]
Hi everyone, 
Since there was no release since 1 year now, how did you guys managed to add it to your projects ?

**Reply by jsvine** [reply]
`pdfminer.six` has, unfortunately, not yet published a release containing that commit. See here: https://github.com/pdfminer/pdfminer.six/blob/master/CHANGELOG.md

**Comment by petermr** [comment]
We are in Geneva presenting our pdfplumber-based pipeline to United Nations
, IPCC, WMO, etc. and there's a lot of interest in converting PDF - they
have metres of reports! So thanks to everyone.

On Fri, Nov 3, 2023 at 6:14 PM Jeremy Singer-Vine ***@***.***>
wrote:

> pdfminer.six has, unfortunately, not yet published a release containing
> that commit. See here:
> https://github.com/pdfminer/pdfminer.six/blob/master/CHANGELOG.md
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/jsvine/pdfplumber/discussions/892#discussioncomment-7469899>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFTCSYM73VBVEQEK4EQUQLYCUYBBAVCNFSM6AAAAAAYQQC6K6VHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM3TINRZHA4TS>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Peter Murray-Rust
Founder ContentMine.org
and
Reader Emeritus in Molecular Informatics
Dept. Of Chemistry, University of Cambridge, CB2 1EW, UK
